component_sdk/python/kfp_component/google/ml_engine/_create_version.py

```diff
@@ -0,0 +1,172 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+import logging
+import time
+
+from googleapiclient import errors
+from fire import decorators
+
+from kfp_component.core import KfpExecutionContext, display
+from ._client import MLEngineClient
+from .. import common as gcp_common
+from ._common_ops import wait_existing_version, wait_for_operation_done
+
+@decorators.SetParseFns(python_version=str, runtime_version=str)
+def create_version(project_id, model_name, deployemnt_uri=None, version_name=None, 
+    runtime_version=None, python_version=None, version=None, 
+    replace_existing=False, wait_interval=30):
+    """Creates a MLEngine version and wait for the operation to be done.
+
+    Args:
+        project_id (str): required, the ID of the parent project.
+        model_name (str): required, the name of the parent model.
+        deployment_uri (str): optional, the Google Cloud Storage location of 
+            the trained model used to create the version.
+        version_name (str): optional, the name of the version. If it is not
+            provided, the operation uses a random name.
+        runtime_version (str): optinal, the Cloud ML Engine runtime version 
+            to use for this deployment. If not set, Cloud ML Engine uses 
+            the default stable version, 1.0. 
+        python_version (str): optinal, the version of Python used in prediction. 
+            If not set, the default version is '2.7'. Python '3.5' is available
+            when runtimeVersion is set to '1.4' and above. Python '2.7' works 
+            with all supported runtime versions.
+        version (str): optional, the payload of the new version.
+        replace_existing (boolean): boolean flag indicates whether to replace 
+            existing version in case of conflict.
+        wait_interval (int): the interval to wait for a long running operation.
+    """
+    if not version:
+        version = {}
+    if deployemnt_uri:
+        version['deploymentUri'] = deployemnt_uri
+    if version_name:
+        version['name'] = version_name
+    if runtime_version:
+        version['runtimeVersion'] = runtime_version
+    if python_version:
+        version['pythonVersion'] = python_version
+
+    return CreateVersionOp(project_id, model_name, version, 
+        replace_existing, wait_interval).execute_and_wait()
+
+class CreateVersionOp:
```

Comment:
Looks like all these Ops sharequite a lot in common, such as execute_and_wait, _dump_metadata, _cancel, etc. 
We can potentially extract a parent class as an interface, wdyt?

Reply:
I am trying to use a composable approach to abstract common code to utilities. If you see the code, there is not too many duplicate code there. It's just the pattern is similar in those files. It's kind of implementation details, we can change it later if there is a better design.

Reply:
Sure, feel free to extract the common interface as the parent class later if we think it is better design.